elfi/examples/daycare.py

```diff
@@ -0,0 +1,301 @@
+
+    The system is solved using the Direct method [Gillespie, 1977].
+
+    References
+    ----------
+    Numminen, E., Cheng, L., Gyllenberg, M. and Corander, J. (2013) Estimating the transmission
+        dynamics of Streptococcus pneumoniae from strain prevalence data, Biometrics, 69, 748-757.
+    Gillespie, D. T. (1977) Exact stochastic simulation of coupled chemical reactions.
+        The Journal of Physical Chemistry 81 (25), 2340–2361.
+
+    Parameters
+    ----------
+    t1 : float or np.array
+        Rate of transmission from other individuals at the DCC.
+    t2 : float or np.array
+        Rate of transmission from the community outside the DCC.
+    t3 : float or np.array
+        Scaling of co-infection for individuals infected with another strain.
+    n_ind : int, optional
+        Number of individuals in a DCC (same for all).
+    n_strains : int, optional
+        Number of bacterial strains considered.
+    freq_strains_commun : np.array of shape (n_strains,), optional
+        Prevalence of each strain in the community outside the DCC.
+    n_obs : int, optional
+        Number of individuals sampled from each DCC (same for all).
+    time_end : float, optional
+        The system is solved using the Direct method until all cases within the batch exceed this.
+    batch_size : int, optional
+    random_state : np.random.RandomState, optional
+
+    Returns
+    -------
+    state_obs : np.array
+        Observations in shape (batch_size, n_dcc, n_obs, n_strains).
+
+    """
+    random_state = random_state or np.random
+
+    t1 = np.asanyarray(t1).reshape((-1, 1, 1, 1))
+    t2 = np.asanyarray(t2).reshape((-1, 1, 1, 1))
+    t3 = np.asanyarray(t3).reshape((-1, 1, 1, 1))
+
+    if freq_strains_commun is None:
+        freq_strains_commun = np.full(n_strains, 0.1)
+
+    prob_commun = t2 * freq_strains_commun
+
+    state = np.zeros((batch_size, n_dcc, n_ind, n_strains), dtype=np.bool)  # infection status
+    time = np.zeros((batch_size, n_dcc))
+    n_factor = 1. / (n_ind - 1)
+    gamma = 1.
+    ind_b_dcc = [np.repeat(np.arange(batch_size), n_dcc), np.tile(np.arange(n_dcc), batch_size)]
+
+    i = 0
+    while np.any(time < time_end):
+        with np.errstate(divide='ignore', invalid='ignore'):
+            # probability of sampling a strain; in paper: E_s(I(t))
+            prob_strain = np.sum(state / np.sum(state, axis=3, keepdims=True),
+                                 axis=2, keepdims=True) * n_factor
+        prob_strain = np.where(np.isfinite(prob_strain), prob_strain, 0)
+
+        # init prob to get infected, same for all
+        hazards = t1 * prob_strain + prob_commun  # shape (batch_size, n_dcc, 1, n_strains)
+
+        # co-infection, depends on the individual's state
+        hazards = np.tile(hazards, (1, 1, n_ind, 1))
+        any_infection = np.any(state, axis=3, keepdims=True)
+        hazards = np.where(any_infection, t3 * hazards, hazards)
+
+        # prob to be cured
```

Comment:
These are not probabilities, yet, but maybe, 'simulated frequencies', or sth?

Reply:
Done